### PR TITLE
Implement HL and Protein feature

### PR DESCRIPTION
**Pull request by cristinamghita**

### CODE REVIEW

File: backend/de.metas.business/src/main/java/de/metas/document/dimension/Dimension.java [hunk 1]
```diff
@@ -111,6 +114,9 @@ public Dimension fallbackTo(@NonNull final Dimension other)
 				.userElementString5(CoalesceUtil.coalesce(this.userElementString5, other.userElementString5))
 				.userElementString6(CoalesceUtil.coalesce(this.userElementString6, other.userElementString6))
 				.userElementString7(CoalesceUtil.coalesce(this.userElementString7, other.userElementString7))
+				.userElement1Id(CoalesceUtil.firstGreaterThanZero(this.userElement1Id, other.userElement1Id))
```
Comment (teosarca) [inline backend/de.metas.business/src/main/java/de/metas/document/dimension/Dimension.java:117]: why not coalese not null?

File: backend/de.metas.adempiere.adempiere/base/src/main/java/de/metas/acct/api/AccountDimension.java [hunk 2]
```diff
@@ -181,6 +182,16 @@ public int getUserElement2_ID()
...
 	}
 
+	public BigDecimal getUserElementNumber1()
+	{
+		return NumberUtils.asBigDecimal(getSegmentValue(AcctSegmentType.UserElementNumber1), BigDecimal.ZERO);
+	}
+
+	public BigDecimal getUserElementNumber2()
+	{
+		return NumberUtils.asBigDecimal(getSegmentValue(AcctSegmentType.UserElementNumber2), BigDecimal.ZERO);
```
Comment (teosarca) [inline backend/de.metas.adempiere.adempiere/base/src/main/java/de/metas/acct/api/AccountDimension.java:192]: why not returning BigDecimal or null ?

File: backend/de.metas.business/src/main/java/de/metas/document/dimension/Dimension.java [hunk 3]
```diff
@@ -194,6 +204,8 @@ else if (collection.size() == 1)
 				.user2_ID(singleElementOrZero(user2_ID))
 				.userElement1Id(singleElementOrZero(userElement1Id))
 				.userElement2Id(singleElementOrZero(userElement2Id))
+				.userElementNumber1(CollectionUtils.singleElementOrDefault(userElementNumber1, BigDecimal.ZERO))
+				.userElementNumber2(CollectionUtils.singleElementOrDefault(userElementNumber2, BigDecimal.ZERO))
```
Comment (teosarca) [inline backend/de.metas.business/src/main/java/de/metas/document/dimension/Dimension.java:208]: why not singleElementOrNull ?
(talking about UserElementNumber1 and UserElementNumber2)

File: backend/de.metas.business/src/main/java/de/metas/order/IOrderLineBL.java [hunk 4]
```diff
@@ -146,7 +147,7 @@ default BigDecimal subtractDiscount(final BigDecimal baseAmount, final BigDecima
 	 * </ul>
 	 */
 	@Nullable
-	I_M_PriceList_Version getPriceListVersion(I_C_OrderLine orderLine);
+	PriceListVersionId getPriceListVersion(I_C_OrderLine orderLine);
```
Comment (teosarca) [inline backend/de.metas.business/src/main/java/de/metas/order/IOrderLineBL.java:150]: pls rename it to getPriceListVersionId

File: backend/de.metas.business/src/main/java/de/metas/order/impl/OrderBL.java [hunk 5]
```diff
@@ -547,7 +548,7 @@ private Optional<BPartnerOrderParams> retrieveBPartnerParams(@NonNull final I_C_
 	}
 
 	@Override
-	public I_M_PriceList_Version getPriceListVersion(final I_C_Order order)
+	public PriceListVersionId getPriceListVersion(final I_C_Order order)
```
Comment (teosarca) [inline backend/de.metas.business/src/main/java/de/metas/order/impl/OrderBL.java:551]: pls rename it to getPriceListVersionId

File: backend/de.metas.business/src/main/java/de/metas/pricing/service/ProductPrices.java [hunk 6]
```diff
@@ -66,29 +66,33 @@
 
 public class ProductPrices
 {
+	final static private IProductDAO productDAO = Services.get(IProductDAO.class);
+	final static private IUOMConversionDAO uomConversionRepo = Services.get(IUOMConversionDAO.class);
+	final static private IPriceListDAO priceListsRepo = Services.get(IPriceListDAO.class);
+	final static private IProductBL productBL = Services.get(IProductBL.class);
```
Comment (teosarca) [inline backend/de.metas.business/src/main/java/de/metas/pricing/service/ProductPrices.java:72]: pls avoid static services!

File: backend/de.metas.ui.web.base/src/main/java/de/metas/ui/web/split_shipment/SplitShipmentRow.java [hunk 7]
```diff
@@ -145,8 +147,8 @@ public SplitShipmentRow withChanges(final List<JSONDocumentChangedEvent> fieldCh
 			{
 				case FIELD_DeliveryDate -> builder.deliveryDate(request.getValueAsLocalDate());
 				case FIELD_QtyToDeliver -> builder.qtyToDeliver(Quantity.of(request.getValueAsBigDecimal(BigDecimal.ZERO), this.qtyToDeliver.getUOM()));
-				case FIELD_UserElementString1 -> builder.userElementString1(request.getValueAsString(null));
-				case FIELD_UserElementString2 -> builder.userElementString2(request.getValueAsString(null));
+				case FIELD_UserElementNumber1 -> builder.userElementNumber1(request.getValueAsBigDecimal(BigDecimal.ZERO));
```
Comment (teosarca) [inline backend/de.metas.ui.web.base/src/main/java/de/metas/ui/web/split_shipment/SplitShipmentRow.java:150]: why can't be null?

File: backend/de.metas.ui.web.base/src/main/java/de/metas/ui/web/split_shipment/SplitShipmentRow.java [hunk 8]
```diff
@@ -118,8 +118,10 @@ public boolean isSaveable()
 	{
 		return deliveryDate != null 
 				&& qtyToDeliver.signum() != 0 
-				&& !Check.isBlank(userElementString1)
-				&& !Check.isBlank(userElementString2);
+				&& userElementNumber1 !=null
+				&& userElementNumber1.signum() != 0
```
Comment (teosarca) [inline backend/de.metas.ui.web.base/src/main/java/de/metas/ui/web/split_shipment/SplitShipmentRow.java:122]: why checking to be non-zero?

File: backend/de.metas.ui.web.base/src/main/java/de/metas/ui/web/split_shipment/SplitShipmentRow.java (hunk 7 above)
Comment (teosarca) [reply backend/de.metas.ui.web.base/src/main/java/de/metas/ui/web/split_shipment/SplitShipmentRow.java:150]: imho zero might be a valid value. we have to support nulls

File: backend/de.metas.contracts/src/main/java/de/metas/contracts/modular/process/ModCntr_Specific_Delete_Price.java [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+	public ProcessPreconditionsResolution checkPreconditionsApplicable(final @NonNull IProcessPreconditionsContext context)
+	{
+		if (!context.isSingleSelection())
+		{
+			return ProcessPreconditionsResolution.rejectBecauseNotSingleSelection();
+		}
+
+		final ModCntrSpecificPriceId contractPriceId = context.getSingleSelectedRecordId(ModCntrSpecificPriceId.class);
+		final ModCntrSpecificPrice contractPrice = modularContractPriceService.getById(contractPriceId);
+		if (!contractPrice.isScalePrice() && NumberUtils.isZeroOrNull(contractPrice.minValue()))
```
Comment (teosarca) [inline backend/de.metas.contracts/src/main/java/de/metas/contracts/modular/process/ModCntr_Specific_Delete_Price.java:50]: contractPrice.isScalePrice might be null => NPE when unboxing

**Review by teosarca — CHANGES_REQUESTED** [review]
check my QA commits & also check my comments

File: backend/de.metas.contracts/src/main/java/de/metas/contracts/modular/computing/purchasecontract/averageonshippedqty/AbstractUENShipmentLineLog.java [hunk 10]
```diff
@@ -150,6 +150,8 @@ public AbstractUENShipmentLineLog(
 				.modularContractTypeId(createLogRequest.getTypeId())
 				.configModuleId(createLogRequest.getConfigId().getModularContractModuleId())
 				.invoicingGroupId(invoicingGroupId)
+				.userElementNumber1(inOutLineRecord.getUserElementNumber1())
```
Comment (teosarca) [inline backend/de.metas.contracts/src/main/java/de/metas/contracts/modular/computing/purchasecontract/averageonshippedqty/AbstractUENShipmentLineLog.java:153]: why not using InterfaceWrapperHelper.getValueAsBigDecimalOrNull?

File: backend/de.metas.business/src/main/java/de/metas/document/dimension/Dimension.java (hunk 1 above)
Comment (teosarca) [inline backend/de.metas.business/src/main/java/de/metas/document/dimension/Dimension.java:117]: i think this shall not be here... u already have it at line 108

File: backend/de.metas.business/src/test/java/de/metas/document/dimension/DimensionTest.java [hunk 11]
```diff
@@ -64,7 +66,8 @@ void fullyPopulated_fallbackTo_empty()
 		void partiallyPopulated_fallbackTo_partiallyPopulated()
 		{
 			final Dimension dim1 = Dimension.builder().productId(ProductId.ofRepoId(1)).build();
-			final Dimension dim2 = Dimension.builder().sectionCodeId(SectionCodeId.ofRepoId(2)).build();
+			final Dimension dim2 = Dimension.builder().sectionCodeId(SectionCodeId.ofRepoId(2))
+					.build();
```
[the remark this answers is not in the document]
Comment (cristinamghita) [reply backend/de.metas.business/src/test/java/de/metas/document/dimension/DimensionTest.java:70]: Nothing
Fixed

File: backend/de.metas.business/src/test/java/de/metas/document/dimension/InvoiceLineDimensionFactoryTest.java [hunk 12]
```diff
@@ -29,6 +31,8 @@ void get_and_set()
 		final Dimension dimensionFromRecordExpected = dimension.toBuilder()
 				.userElement1Id(0)
 				.userElement2Id(0)
+				.userElementNumber1(BigDecimal.ZERO)
```
Comment (teosarca) [inline backend/de.metas.business/src/test/java/de/metas/document/dimension/InvoiceLineDimensionFactoryTest.java:34]: nop! those shall be null!

Comment (teosarca) [reply backend/de.metas.business/src/test/java/de/metas/document/dimension/InvoiceLineDimensionFactoryTest.java:34]: fix ur de.metas.document.dimension.InvoiceLineDimensionFactory#getFromRecord implementation

File: backend/de.metas.cucumber/src/test/resources/de/metas/cucumber/features/groupInvoiceCandidates.feature [hunk 13]
```diff
@@ -1,3 +1,4 @@
+@dev:runThisOne
```
Comment (teosarca) [inline backend/de.metas.cucumber/src/test/resources/de/metas/cucumber/features/groupInvoiceCandidates.feature:1]: i think u don't need this anymore?

File: backend/de.metas.swat/de.metas.swat.base/src/test/java/de/metas/invoicecandidate/document/dimension/InvoiceCandidateDimensionFactoryTest.java [hunk 14]
```diff
@@ -34,6 +36,8 @@ void get_and_set()
 				.user2_ID(0)
 				.userElement1Id(0)
 				.userElement2Id(0)
+				.userElementNumber1(BigDecimal.ZERO)
+				.userElementNumber2(BigDecimal.ZERO)
```
Comment (teosarca) [inline backend/de.metas.swat/de.metas.swat.base/src/test/java/de/metas/invoicecandidate/document/dimension/InvoiceCandidateDimensionFactoryTest.java:40]: i think it shall be null!

**Review by teosarca — CHANGES_REQUESTED** [review]
check my comments

File: backend/de.metas.contracts/src/main/sql/postgresql/ddl/views/ModCntr_UseElementNumber_V.sql [hunk 15]
```diff
@@ -0,0 +1,27 @@
+DROP VIEW ModCntr_UseElementNumber_V;
+
+CREATE OR REPLACE VIEW ModCntr_UseElementNumber_V AS
```
[the remark this answers is not in the document]
Comment (cristinamghita) [reply backend/de.metas.contracts/src/main/sql/postgresql/ddl/views/ModCntr_UseElementNumber_V.sql:3]: we need for reports

File: backend/de.metas.business/src/test/java/de/metas/document/dimension/InvoiceLineDimensionFactoryTest.java [hunk 16]
```diff
@@ -6,6 +6,8 @@
 import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Test;
 
+import java.math.BigDecimal;
```
Comment (teosarca) [inline backend/de.metas.business/src/test/java/de/metas/document/dimension/InvoiceLineDimensionFactoryTest.java:9]: unnecessary import

**Review by teosarca — APPROVED** [review]

**Pull request merged**